### PR TITLE
fix(49838): "Extract function" refactoring action is disabled for a wrong reason

### LINKED ISSUE #49838: "Extract function" refactoring action is disabled for a wrong reason
# Bug Report
Selecting a piece of code that contains a loop and a conditional break would disable "Extract function" refactoring action even if the whole loop is inside the selection.

### 🔎 Search Terms
"conditional break"
"Cannot extract range"

### 🕗 Version & Regression Information
Tried on 4.7.3 and 4.7.4
I just noticed it now. Didn't try it before.

### ⏯ Playground Link
[Playground link with relevant code](https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBAMwK4DsDGMCWEVwLYCeAYqmgBQRhY4BccKSeARsFAJR0POtwA+9SADaC4AbzgAoAJBocAZ3h4IAE2BwAvHErUUAPnUAGOAH44AIgBKAQxTKIeM3DpmimEMGVmA3BOmDg8FDAckIwnIwsUHwCwhoxgj5SmAhwZEqqGuqaljZ2Dmxi0lIA9MVwAMrA-hjYuApWsEUI0Kn+8JhxBl5wHQA8cACMXT0A1CMFokUy8oG5cQCyVjAAFgB0ULn2ZGyJUkkpZBu2cLpwBqsArBNTUkEhgvCaR8q7e0xBVgDWrwC+RX97UoVKrAGo4ODAWzSAFVORqcRFO6hTqJP7SIIwJBQXBIh6JCQ-IA)

### 💻 Code

```ts
export function myFunc(option: number): number | null { 
	const mode = option>=0 ? "Random" : "Fixed";

	let result: number | null = null;
	if (mode === "Random") {
		// Selection start
		for (let i = 0; i < 10; i++) {
			const rand = Math.random();
			if (rand > 0.5) {
				result = rand;
				break; // This causes the refactoring to be disabled
			}
		}
		// Selection end
	}
	else { 
		result = 0;
	}

	return result;
	
}
```

### 🙁 Actual behavior
``` json
{
	"name": "Extract Symbol",
	"description": "Extract function",
	"actions": [
		{
			"name": "Extract Function",
			"description": "Extract function",
			"kind": "refactor.extract.function",
			"notApplicableReason": "Cannot extract range containing conditional break or continue statements."
		}
	]
},
```

### 🙂 Expected behavior
Extract function should be allowed as the loop is within the Selection

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #49838

### CODE REVIEW

File: src/services/refactors/extractSymbol.ts
Comment (amcasey) [reply src/services/refactors/extractSymbol.ts:555]: Oh, I see - it's because we can only handle unconditional jumps (unless the jump target is in the range, of course).

Comment (amcasey) [reply src/services/refactors/extractSymbol.ts:555]: So I think we want

```ts
case SyntaxKind.IfStatement:
    permittedJumps &= ~PermittedJumps.Return;
    break;
```

**Review by amcasey — CHANGES_REQUESTED** [review]
The fix looks correct but incomplete.

File: src/services/refactors/extractSymbol.ts
Comment (a-tarasyuk) [reply src/services/refactors/extractSymbol.ts]: @amcasey Thanks for the review. I tried to use this fix, however, I'm unsure if we should allow extraction for the following tests.

https://github.com/microsoft/TypeScript/blob/298b3a432ca099330108ed7bb18293ae79693672/src/testRunner/unittests/services/extract/ranges.ts#L213-L245

Comment (amcasey) [reply src/services/refactors/extractSymbol.ts:555]: No, those extractions shouldn't be allowed.  I thought we started permitting `None`, so I wasn't expecting anything to make `Break` permissible in that scenario.  I'll have to have another look at the code.

Comment (a-tarasyuk) [reply src/services/refactors/extractSymbol.ts:555]: @amcasey Oke, thx.

Comment (amcasey) [reply src/services/refactors/extractSymbol.ts:555]: @a-tarasyuk Are you seeing those tests fail?  They're passing for me with my proposed change (or I'm doing something dumb).

**Review by amcasey — APPROVED** [review]
Tests are green and the change LGTM, but I wouldn't complain if you added a test for the switch case that the updated fix handles. :wink:

**Comment by a-tarasyuk** [comment]
@amcasey Do you mean the following test?
```ts
export function fn(m: number, n: number) {
   const mode = m >= 0 ? "a" : "b";
   let result: number = 0;

   if (mode === "a") {
       /*a*/for (let i = 0; i < n; i++) {
           const rand = Math.random();
           switch (rand) {
                case 0.1:
                    result = rand;
                break;
                default:
                    result = 1;
                break;
           }
       }/*b*/
   }
   else {
       result = 0;
   }

   return result;
}

```

**Comment by a-tarasyuk** [comment]
@amcasey I've added that test too.

**Review by amcasey — APPROVED** [review]
Thanks!

**Comment by amcasey** [comment]
Oh, I just realized you probably don't have merge permissions. 🤦‍♂️

**Comment by a-tarasyuk** [comment]
@amcasey Yes, I don't., That's ok because I'm not an MS employee 😄

**Comment by amcasey** [comment]
@a-tarasyuk Could have fooled me.  :wink:  I'm just so used to signing off on things on DefinitelyTyped and having the author & bot sort things out.

**Pull request merged**
